src/dune_rules/findlib/findlib.ml

```diff
@@ -97,44 +104,161 @@ module Vars = struct
+      ; ("ocamlpath", Dyn.list Path.to_dyn ocamlpath)
+      ; ("toolchain", option string toolchain)
       ]
 
-  let load path ~toolchain ~context =
-    let path = Path.Outside_build_dir.extend_basename path ~suffix:".d" in
-    let conf_file =
-      Path.Outside_build_dir.relative path (toolchain ^ ".conf")
-    in
-    let* conf_file_exists = Fs_memo.file_exists conf_file in
-    if not conf_file_exists then
-      User_error.raise
-        [ Pp.textf "ocamlfind toolchain %s isn't defined in %s (context: %s)"
-            toolchain
-            (Path.Outside_build_dir.to_string_maybe_quoted path)
-            context
-        ];
-    let+ meta = Meta.load ~name:None conf_file in
-    { vars = String.Map.map meta.vars ~f:Rules.of_meta_rules
-    ; preds = Ps.of_list [ P.make toolchain ]
-    }
+  let ocamlpath_sep =
+    if Sys.cygwin then (* because that's what ocamlfind expects *)
+      ';'
+    else Bin.path_sep
 
-  let get { vars; preds } var = Vars.get vars var preds
+  let path_var = Bin.parse_path ~sep:ocamlpath_sep
 
-  let env t =
-    let preds = Ps.add t.preds (P.make "env") in
-    String.Map.filter_map ~f:(Rules.interpret ~preds) t.vars
+  let ocamlpath env = Env.get env "OCAMLPATH" |> Option.map ~f:path_var
+
+  let set_toolchain t ~toolchain =
+    match t.toolchain with
+    | None ->
+      { t with
+        config = File.toolchain t.config ~toolchain
+      ; toolchain = Some toolchain
+      }
+    | Some old_toolchain ->
+      Code_error.raise
+        "Findlib.Config.set_toolchain: cannot set toolchain twice"
+        [ ("old_toolchain", Dyn.string old_toolchain)
+        ; ("toolchain", Dyn.string toolchain)
+        ]
+
+  let path t =
+    match File.get t.config "path" with
+    | None -> t.ocamlpath
+    | Some p -> t.ocamlpath @ path_var p
+
+  let tool t ~prog =
+    match File.get t.config prog with
+    | None -> Memo.return None
+    | Some s -> (
+      match Filename.analyze_program_name s with
+      | In_path -> t.which s
+      | Relative_to_current_dir ->
+        User_error.raise
+          [ Pp.textf
+              "The effective Findlib configuration specifies the relative path \
+               %S for the program %S. This is currently not supported."
+              s prog
+          ]
+      | Absolute ->
+        Memo.return (Some (Path.of_filename_relative_to_initial_cwd s)))
+
+  let ocamlfind_config_path ~env ~which ~findlib_toolchain =
+    let open Memo.O in
+    let+ path =
+      match Env.get env "OCAMLFIND_CONF" with
+      | Some s -> Memo.return (Some s)
+      | None -> (
+        match findlib_toolchain with
+        | None -> Memo.return None
+        | Some _ -> (
+          which "ocamlfind" >>= function
+          | None -> Memo.return None
+          | Some fn ->
+            Memo.of_reproducible_fiber
+              (Process.run_capture_line ~display:Quiet ~env Strict fn
+                 [ "printconf"; "conf" ])
+            |> Memo.map ~f:Option.some))
```

Comment:
Personally, I find it weird to see code like `g (f x) |> h`. It might as well be `f x |> g |> h`